### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] Apparent false positive when type parameter is used once in function signature but is passed into a generic function signature

### ISSUE THREAD

**Issue by BernardoFuret** [typescript-eslint/typescript-eslint#11528]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgJIFEpQPZQDwAqAInGHAHzIDeAsAFDLLQ5QBcyYUArhANz2MAJqTjtiI-nQC%2B9egmwgAzmGQIAFhAQBrVIswtkAXmSESZADTkAFHBABPMGtABzdlxBaQ2AO4gAlOy2Dk4gzsjAimj6uKYilIaUtAzIUBBgXFAgyFYCHHYADhDYMMhBji5GhsYA5NgARgBWmmDVyABkbbkAhF1lIWEdudXMuK2gpfbloe2dyWAFRSV9LgB0I1CVNXXY2AA2ELatg8nLoWtYuLl%2BklKScgrKTAAenIhg0VBmcEYm4hbW6wA-G4PF5fAFkH9vgAfZAgLi7XZGRK5VLpTKqDTaXQfWJkAEXKB%2BZCApiElbCMjIdjwxE3XhAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUFtQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
interface IError<TData> {
  error: true;
  data: TData;
}

const checkIsError = <TData>(anything: unknown): anything is IError<TData> => {
  return (
    typeof anything === 'object' &&
    !!anything &&
    'error' in anything &&
    typeof anything.error === 'boolean' &&
    anything.error
  );
};

const extractErrorData = <TData>(error?: unknown): TData | null => {
  return checkIsError<TData>(error) ? error.data : null;
};
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I didn't expect the error on `TData`, because it is passed to the type guard and is needed to provide the return type of the `extractErrorData` function.

### Actual Result

I receive an error on the generic `TData` of `extractErrorData` saying `Type parameter TData is used only once in the function signature`.

### Additional Info

Hi, 
I read the docs and I tried to search for a similar case, but didn't find any answers. My apologies if this has been covered or falls into a described case and I simply failed to understand it.

Now onto the problem:
I understand the suggestion to replace all usages of the type parameter with its constraint (`unknown`, in this case), but that would mean I can't type the `data` field of the object. I understand there's no runtime guarantee of the structure of the `data` field (as in: I'm not performing any checks on the structure of the `data` field). But from a type descriptions perspective, shouldn't it work, just like type guards do?

Then, why does the following work?
```ts
const extractErrorData = <TData,>(error?: unknown): IError<TData> | null => {
  return checkIsError<TData>(error) ? error : null;
};
```
Here, the returned type is `IError<TData>` instead of `TData`. The same checks are applied to `TData`, but there's no alert from the rule.
([Here's a playground link](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgJIFEpQPZQDwAqAInGHAHzIDeAsAFDLLQ5QBcyYUArhANz2MAJqTjtiI-nQC%2B9egmwgAzmGQIAFhAQBrVIswtkAXmSESZADTkAFHBABPMGtABzdlxBaQ2AO4gAlOy2Dk4gzsjAimj6uKYilIaUtAzIUBBgXFAgyFYCHHYADhDYMMhBji5GhsYA5NgARgBWmmDVyABkbbkAhF1lIWEdudXMuK2gpfbloe2dyWAFRSV9LgB0I1CVNXXY2AA2ELatg8nLoWtYuLl%2BklKScgrKTAAenIhg0VBmcEYm4hbW6wA-G4PF5fAEohd8H8KMgAD7IEBcXa7IyJXKpdKZVQabS6D6xMgAqF%2BZCAphQ5DsJEom68IA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUFtQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

Should the rule stop alerting on my original case? Or should the rule also alert for this second case? Or am I seeing this wrong?

Any help is greatly appreciated, thank you!

**Comment by bradzacher** [comment]
This is working as expected.
Your code is only using the generic exactly once -- in the return type.
In a nutshell a generic must appear in 2 or more locations in the signature for it to to not be unnecessary.

Especially in this case. You have only used the generic in the return type which is unnecessary and also highly unsafe! Generics that are _only_ used in the return type create a false sense of security because it can make functions appear like they're returning one type when they're returning a different type

Your code could be rewritten without the generic to use `unknown` and work exactly the same:

```ts
interface IError<TData> {
  error: true;
  data: TData;
}

const checkIsError = (anything: unknown): anything is IError<unknown> => {
  return (
  typeof anything === 'object' &&
  !!anything &&
  'error' in anything &&
  typeof anything.error === 'boolean' &&
  anything.error
  );
};

const extractErrorData = (error?: unknown): unknown | null => {
  return checkIsError(error) ? error.data : null;
};
```

You might say "but without the generic I can't change the type of `TData` in `IError` and that's correct. But that is the correct and _safe_ behaviour. In order to change the type of the generic one would need to use an assertion -- `extractErrorData(error) as IError<MyType> | null` -- which clearly shows the unsafety of the operation in the code.

**Comment by BernardoFuret** [comment]
Hey,
Thank you for your explanation!

Hope it's fine to leave a message here, as it's still not clear to me why this was closed.

I had already understood the lack of safety on `TData` from my example. But I don't understand why the linter doesn't flag this case:
```ts
const extractErrorData = <TData,>(error?: unknown): IError<TData> | null => {
  return checkIsError<TData>(error) ? error : null;
};
```

Shouldn't it also flag this case (all else equal)? `extractErrorData` will possibly return `IError<TData>`, where `TData` is as `unknown` as in my original example. It's unsafe.

The same argument applies. It should be written as:
```ts
const extractErrorData = (error?: unknown): IError<unknown> | null => {
  return checkIsError<unknown>(error) ? error : null;
};
```

Cheers!
